src/charm.py

```diff
@@ -633,7 +669,66 @@ def _get_backup_key(self) -> str:
         timestamp = datetime.datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
         return f"{BACKUP_KEY_PREFIX}-{self.model.name}-{timestamp}"
 
-    def _get_ca_cert_location_in_charm(self) -> str:
+    def _on_new_access_certificate_joined(self, event: RelationJoinedEvent):
+        if not self._ingress_address:
+            raise RuntimeError("Tried to process relation join before peer relation.")
+        if not (private_key := self._get_private_key_from_workload().encode()):
+            private_key = generate_private_key()
+        new_csr = generate_csr(private_key=private_key, subject=self._ingress_address)
+        self._push_csr_to_workload(new_csr.decode())
+        self._push_unit_pk_to_workload(private_key=private_key.decode())
+        self.tls_certificate_access.request_certificate_creation(
+            certificate_signing_request=new_csr
+        )
+        logger.info(
+            "New certificate requested from %s for unit %s",
+            event.relation.name,
+            self.unit.name,
+        )
+
+    def _on_new_access_certificate_available(self, event: CertificateAvailableEvent):
+        private_key = self._get_private_key_from_workload()
+        csr = self._get_csr_from_workload()
+        if not private_key or csr != event.certificate_signing_request:
+            return
+        self._push_unit_certificate_to_workload(certificate=event.certificate)
+        self._push_ca_certificate_to_workload(certificate=event.ca)
+        self._container.send_signal(SIGHUP, self._container_name)
+
+    def _on_access_certificate_expiring(self, event: CertificateExpiringEvent):
+        private_key = self._get_private_key_from_workload()
+        old_csr = self._get_csr_from_workload()
+        old_cert = self._get_cert_from_workload()
+        if event.certificate != old_cert:
+            return
+        if not self._ingress_address:
+            raise RuntimeError("Peer relation somehow unavailable. Something is very wrong")
+        new_csr = generate_csr(private_key=private_key.encode(), subject=self._ingress_address)
+        self._push_csr_to_workload(new_csr.decode())
+        self.tls_certificate_access.request_certificate_renewal(
+            old_certificate_signing_request=old_csr.encode(),
+            new_certificate_signing_request=new_csr,
+        )
+        logger.info("Sent a certificate renewal request for %s.", self.unit.name)
+
+    def _on_access_certificate_invalidated(
+        self, event: Union[AllCertificatesInvalidatedEvent, CertificateInvalidatedEvent]
+    ):
+        if (
+            hasattr(event, "certificate_signing_request")
+            and event.certificate_signing_request != self._get_csr_from_workload()
+        ):
+            return
+        self._remove_all_certs_from_workload()
+        self._restore_self_signed_certs_to_workload()
+        self._container.send_signal(SIGHUP, self._container_name)
```

Comment:
Why use SIGHUP instead of stopping the service with Pebble?

Reply:
I assume `SIGHUP` is used to reload the configuration without stopping in this case.

Reply:
The documentation also explicitly mentions using SIGHUP to reload certificates [here](https://developer.hashicorp.com/vault/docs/configuration/listener/tcp). in `tls_cert_file` and `tls_key_file`.

Reply:
There are also multiple benefits to using this. First is that there is no downtime. Any request that comes while the process is reloading is queued, so the certificate switchover appears instant.

This also means that if for example a config change happens, it's not possible to catch the container in an unresponsive state which could incorrectly put the unit back to a waiting state. Handling this issue in charm code is a little difficult without having the ability to use defer() for example.

I think it makes more sense to ask why should we restart when we can use `SIGHUP` instead of asking why should we use `SIGHUP` when we can restart.